_posts/2020-08-06-textpaint-measure.md

```diff
@@ -0,0 +1,108 @@
+
+The method requires several values:
+- `text` is the text we want to measure,
+- `start` is where in the `text` we should start measuring,
+- `end` is where in the `text` we should stop measuring,
+- `bounds` is where the results of the measurement will be stored
+
+The Javadoc says that the caller (me, Zarah) should allocate the [`Rect`](https://developer.android.com/reference/kotlin/android/graphics/Rect?hl=en).
+```kotlin
+val textBounds = Rect()
+```
+
+We can then give this `Rect` to `getTextBounds()`:
+```kotlin
+textPaint.getTextBounds(text, 0, 3, textBounds)
+```
+
+And now our `textBounds` will contain a bunch of values. Remember that the font and the font size affects how much space we need to draw our text, and this is what I got for a custom font and size of 10sp:
+```
+Rect(
+    bottom = 0
+    left = 2
+    right = 66
+    top = -20
+    )
+```
+
+Turned into a photo:
+<center>
+    <a href="https://imgur.com/rsALySp"><img src="https://i.imgur.com/rsALySp.jpg" title="source: imgur.com" /></a>
+</center>
+
+The `"implied origin at (0,0)"` the Javadoc refers to is the leftmost edge of the baseline and the values in our `textBounds` are relative to this point (In contrast to the `Canvas` where `(0, 0)` is the top-left). Since our text is all caps, the `bottom` coordinate of our bounding box aligns with the baseline (how convenient). Values in `textBounds` **increase** as you go down and to the right of the origin, and **decrease** as you go the opposite way.
+
+<center>
+    <a href="https://imgur.com/T65MifE"><img src="https://i.imgur.com/T65MifE.jpg" title="source: imgur.com" /></a>
+    <br /><small>Convenience</small>
+</center>
+
+It is important to note here that the text _may not be_ drawn exactly over the origin -- there _may be_ a leading space between the first letter and the leftmost edge of our bounding box. (I don't know enough about typography but from my experiments it looks like this depends on how the font renders each letter).
+
+This is all fine and good, but what if we give it "`Something`" else? We now have a mix of lower- and uppercase letters, as well as letters with _ascenders_ (things that go up) and _descenders_ (things that go down -- but in my head I call them tails).
+
+With the same custom font and size, our `textBounds` now look like this:
+```
+Rect(
+    bottom = 6
+    left = 1
+    right = 133
+    top = -20
+    )
+```
+
+Turned into a photo:
+<center>
+    <a href="https://imgur.com/inOazdv"><img src="https://i.imgur.com/inOazdv.jpg" title="source: imgur.com" /></a>
+</center>
+
+Now that we have enough information about how far down from the baseline our descent line is and how far up from the baseline our ascent line is (look at us using typography terms! Go us!), we can do the necessary maths to tell Android where exactly we want our text to be drawn.
```

Comment:
"we can do the necessary maths" - here i was expecting to see the math you need to do.

I like that you're starting from a practical problem, but without the math it doesn't seem that you solved the initial problem